### DISCUSSION TITLE
[CONTRACTS] - Stabilization Automation

### DISCUSSION THREAD

**Discussion by 0x4007** [pavlovcik/uad-ui#140] [Ideas]
### Definitions

`collateral treasury` - where we store the Proxy Yield Aggregator yields.

### On-Swap Top-Up Strategy
Leveraging transfer hooks, we:
1. monitor the uAD-3CRV market for sells below 1.00 TWAP 
2. automatically deposit single sided 3CRV liquidity from the collateral treasury.
### Optimization (Removal of Single Sided Liquidity)
I wonder if we should remove some uAD liquidity to make it less burdensome on our collateral treasury balance? 

A. Should this ratio be 1:1 e.g. we remove 1 uAD and add 1 3CRV during this topup? 
B. Should we make this ratio dynamic based on the amount of 3CRV market liquidity compared to collateral treasury?
  * `100000 'extra' uAD in market` ('extra' meaning that if we remove all of this, then the price will be 1.00) : `1000 3CRV in collateral treasury` calculation is (`100000/100=100`) yields withdrawal of `100 uAD` and deposit of `1 3CRV`?

C. What's the optimal formula?

This way, instead of a burn disincentive there is just a higher gas fee disincentive to sell. 

### Further Ramblings

* We should have the collateral treasury ideally be put to work. Should it just be in our Proxy Yield Aggregator or should we leave this capital idle?
* Imagine whitelisting other markets and having this just-in-time top-up tech exist across all major markets to truly have an enforced pegged stablecoin across the entire ecosystem!

**Comment by hpiyankov** [comment]
This can work, but gut feeling says it can be exploitable, I just need to figure out how.

The question is, do you need this researched now? Shall we first get treasury/traction before figuring out how to spend it :)

**Comment by 0x4007** [comment]
![Screenshot_20211112-082635_Chrome](https://user-images.githubusercontent.com/4977000/141420627-9ea2dfd8-ad05-461e-8557-950617c256e1.jpg)
Context: these are the stabilization mechanics, in order of "line of defense" taken from our pitchdeck. Basically we should intelligently combine the addition of collateral+removal of SSL 
![Uploading image.png…]()
